tls/s2n_resume.c

```diff
@@ -153,10 +154,32 @@ static int s2n_tls12_deserialize_resumption_state(struct s2n_connection *conn, s
     S2N_ERROR_IF(then > now, S2N_ERR_INVALID_SERIALIZED_SESSION_STATE);
     S2N_ERROR_IF(now - then > conn->config->session_state_lifetime_in_nanos, S2N_ERR_INVALID_SERIALIZED_SESSION_STATE);
 
-    /* Last but not least, put the master secret in place */
     POSIX_GUARD(s2n_stuffer_read_bytes(from, conn->secrets.master_secret, S2N_TLS_SECRET_LEN));
 
-    return 0;
+    /* TODO: https://github.com/aws/s2n-tls/issues/2990 */
+    if (S2N_IN_TEST && s2n_stuffer_data_available(from)) {
+        uint8_t ems_negotiated = 0;
+        POSIX_GUARD(s2n_stuffer_read_uint8(from, &ems_negotiated));
+
+        /**
+         *= https://tools.ietf.org/rfc/rfc7627#section-5.3
+         *# If the original session did not use the "extended_master_secret"
+         *# extension but the new ClientHello contains the extension, then the
+         *# server MUST NOT perform the abbreviated handshake.  Instead, it
+         *# SHOULD continue with a full handshake (as described in
+         *# Section 5.2) to negotiate a new session.
+         *#
+         *# If the original session used the "extended_master_secret"
+         *# extension but the new ClientHello does not contain it, the server
+         *# MUST abort the abbreviated handshake.
+         **/
+        if (conn->ems_negotiated != ems_negotiated) {
+            conn->ems_negotiated = ems_negotiated;
+            POSIX_BAIL(S2N_ERR_INVALID_SERIALIZED_SESSION_STATE);
```

Comment:
Talked offline: this is a little odd, and might be a good use case for a second flag. But the second flag could also make the check in s2n_conn_set_handshake_type more confusing. Probably not a huge difference in readability.

Reply:
I added a comment to clarify what's going on. TBH I'm unsure if having a second flag will help here.